File: src/game/playerSmashSkill.ts
import type { CharacterMoveDirection } from './characterState'

export const PLAYER_SMASH_SKILL_ID = 'smash'
export const PLAYER_SMASH_SKILL_SEGMENT_COUNT = 3
export const PLAYER_SMASH_SKILL_FIRST_SEGMENT_DISTANCE_PIXELS = 22
export const PLAYER_SMASH_SKILL_SEGMENT_SPACING_PIXELS = 14
export const PLAYER_SMASH_SKILL_LAUNCH_OFFSET_PIXELS = 28
export const PLAYER_SMASH_SKILL_TRAVEL_DISTANCE_PIXELS = 96
export const PLAYER_SMASH_SKILL_SEGMENT_STAGGER_MILLISECONDS = 45
export const PLAYER_SMASH_SKILL_SEGMENT_DURATION_MILLISECONDS = 360
export const PLAYER_SMASH_SKILL_COOLDOWN_MILLISECONDS = 300
export const PLAYER_SMASH_SKILL_EFFECT_ANIMATION_SPEED = 0.38
export const PLAYER_SMASH_SKILL_EFFECT_BASE_SCALE_X = 0.3
export const PLAYER_SMASH_SKILL_EFFECT_BASE_SCALE_Y = 0.26
export const PLAYER_SMASH_SKILL_EFFECT_SCALE_STEP = 0.025
export const PLAYER_SMASH_SKILL_EFFECT_SCALE_GROWTH = 0.08

export type PlayerSmashSkillSegmentPlacement = {
  x: number
  y: number
  scaleX: number
  scaleY: number
  alpha: number
}

export const getPlayerSmashSkillDirectionVector = (
  facing: CharacterMoveDirection
): {
  x: number
  y: number
} => {
  switch (facing) {
    case 'up':
      return { x: 0, y: -1 }
    case 'down':
      return { x: 0, y: 1 }
    case 'left':
      return { x: -1, y: 0 }
    case 'right':
      return { x: 1, y: 0 }
    default:
      return { x: 0, y: 0 }
  }
}

export const getPlayerSmashSkillSegmentDistancePixels = (
  segmentIndex: number,
  progress = 0
): number =>
  PLAYER_SMASH_SKILL_LAUNCH_OFFSET_PIXELS +
  PLAYER_SMASH_SKILL_FIRST_SEGMENT_DISTANCE_PIXELS +
  segmentIndex * PLAYER_SMASH_SKILL_SEGMENT_SPACING_PIXELS +
  (1 - (1 - Math.min(1, Math.max(0, progress))) ** 2) *
    PLAYER_SMASH_SKILL_TRAVEL_DISTANCE_PIXELS

export const getPlayerSmashSkillSegmentPlacement = ({
  characterCenterX,
  characterCenterY,
  facing,
  segmentIndex,
  progress = 0
}: {
  characterCenterX: number
  characterCenterY: number
  facing: CharacterMoveDirection
  segmentIndex: number
  progress?: number
}): PlayerSmashSkillSegmentPlacement => {
  const direction = getPlayerSmashSkillDirectionVector(facing)
  const normalizedProgress = Math.min(1, Math.max(0, progress))
  const distance = getPlayerSmashSkillSegmentDistancePixels(
    segmentIndex,
    normalizedProgress
  )
  const pulseProgress = Math.sin(normalizedProgress * Math.PI)
  const growthMultiplier =
    1 + pulseProgress * PLAYER_SMASH_SKILL_EFFECT_SCALE_GROWTH
  const baseScaleX =
    PLAYER_SMASH_SKILL_EFFECT_BASE_SCALE_X +
    segmentIndex * PLAYER_SMASH_SKILL_EFFECT_SCALE_STEP
  const baseScaleY =
    PLAYER_SMASH_SKILL_EFFECT_BASE_SCALE_Y +
    segmentIndex * PLAYER_SMASH_SKILL_EFFECT_SCALE_STEP
  const fadeOutStartProgress = 0.78
  const alpha =
    normalizedProgress < fadeOutStartProgress
      ? 1
      : Math.max(
          0,
          1 -
            (normalizedProgress - fadeOutStartProgress) /
              (1 - fadeOutStartProgress)
        )

  return {
    x: characterCenterX + direction.x * distance,
    y: characterCenterY + direction.y * distance,
    scaleX: baseScaleX * growthMultiplier,
    scaleY: baseScaleY * growthMultiplier,
    alpha
  }
}

File: src/games/my-sample-rpg/assets/lua/player-smash-skill.lua
-- TS playerSmashSkill.ts → Lua. 스매시 스킬의 세그먼트 배치 수학(방향 벡터 / 거리 / 배치).
-- 순수 수학이라 상수는 TS에서 인자로 전달(상수 소유는 TS, 공식만 Lua). 객체 입출력이라 JSON 마샬링.
-- facing 방향 스위치, 거리(min/max + 정수 제곱), 배치(sin(progress*pi)).

local function clamp01(value)
  if value < 0 then
    return 0
  end
  if value > 1 then
    return 1
  end
  return value
end

-- facing -> { x, y } 단위 벡터. (TS의 switch 기본값은 {0,0}.)
function player_smash_skill_direction_vector(facing)
  if facing == 'up' then
    return { x = 0, y = -1 }
  elseif facing == 'down' then
    return { x = 0, y = 1 }
  elseif facing == 'left' then
    return { x = -1, y = 0 }
  elseif facing == 'right' then
    return { x = 1, y = 0 }
  end
  return { x = 0, y = 0 }
end

-- 세그먼트 거리(px). progress는 0..1로 클램프, (1-(1-p)^2) 이징 곡선.
-- TS의 ** 2 는 double 곱이라 (1-p)*(1-p) 와 비트 동일.
function player_smash_skill_segment_distance_pixels(
  segment_index,
  progress,
  launch_offset,
  first_segment_distance,
  segment_spacing,
  travel_distance
)
  local normalized = clamp01(progress)
  local inverse = 1 - normalized
  return launch_offset
    + first_segment_distance
    + segment_index * segment_spacing
    + (1 - inverse * inverse) * travel_distance
end

-- 세그먼트 배치 결과 { x, y, scaleX, scaleY, alpha }.
function player_smash_skill_segment_placement(input, constants)
  local direction = player_smash_skill_direction_vector(input.facing)
  local normalized = clamp01(input.progress)
  local distance = player_smash_skill_segment_distance_pixels(
    input.segmentIndex,
    normalized,
    constants.launchOffset,
    constants.firstSegmentDistance,
    constants.segmentSpacing,
    constants.travelDistance
  )
  local pulse = math.sin(normalized * math.pi)
  local growth_multiplier = 1 + pulse * constants.scaleGrowth
  local base_scale_x = constants.baseScaleX + input.segmentIndex * constants.scaleStep
  local base_scale_y = constants.baseScaleY + input.segmentIndex * constants.scaleStep
  local fade_out_start = 0.78
  local alpha
  if normalized < fade_out_start then
    alpha = 1
  else
    local faded = 1 - (normalized - fade_out_start) / (1 - fade_out_start)
    alpha = faded > 0 and faded or 0
  end

  return {
    x = input.characterCenterX + direction.x * distance,
    y = input.characterCenterY + direction.y * distance,
    scaleX = base_scale_x * growth_multiplier,
    scaleY = base_scale_y * growth_multiplier,
    alpha = alpha
  }
end
